three algorithms that are widely used for building
recommendation systems:
1. Association Rules
2. Collaborative Filtering
3. Matrix Factorization

ASSOCIATION RULES (ASSOCIATION RULE MINING)

Association rule finds combinations of items that frequently occur together in orders or baskets.
An application of association rule mining is in Market Basket Analysis (MBA).

Association rule considers all possible combination of items in the previous baskets and computes various measures such as support, confidence, and lift to identify rules with stronger associations.

Depending on the possible combination, association rule mining may require huge computational power to go through all possible combinations.

One solution to this problem is to eliminate items that possibly cannot be part of any itemsets.

One such algorithm the association rules use apriori algorithm. The apriori algorithm was proposed by Agrawal and Srikant (1994). The rules generated are represented as {diapers} -> {beer} which means that customers who purchased diapers also purchased beer in the same basket. {diaper,beer} together is called itemset. 

{diaper} is called the antecedent and the {beer} is called the consequent.

Both antecedents and consequents can have multiple items, e.g. {diaper, milk} -> {beer, bread} is also a valid rule.

Metrics
Concepts such as support, confidence, and lift are used to generate association rules.


1. Support
Support indicates the frequencies of items appearing together in baskets with respect to all possible baskets being considered (or in a sample). For example, the support for (beer, diaper) will be 2/4 or 50% as it appears together in 2 baskets out of 4 baskets.

1. N be the total number of baskets.
2. NXY represent the number of baskets in which X and Y appear together.
3. NX represent the number of baskets in which X appears.
4. NY represent the number of baskets in which Y appears.
Then the support between X and Y, Support(X, Y)= NXY/N

To filter out stronger associations, we can set a minimum support (for example, minimum support of 0.01).
This means the itemset must be present in at least 1% of baskets. Apriori algorithm uses minimum support criteria to reduce the number of possible itemset combinations, which in turn reduces computational requirements.

Hence, apriori algorithm computes support for each item independently and eliminates items with support less than minimum support.


2. Confidence
Confidence measures the proportion of the transactions that contain X, which also contain Y. X is called antecedent and Y is called consequent. Confidence can be calculated using the following formula:

Confidence(X -> Y)= P (Y | X)= NXY/NX

where P(Y|X) is the conditional probability of Y given X.

3. Lift 
It is calculated using the following formula:

Lift = Support (X,Y)/(Support(X)*Support(Y))=NXY/(NX*NY)

Lift can be interpreted as the degree of association between two items. 

> Lift value 1 indicates that the items are independent (no association), lift value of less than 1 implies that the products are substitution (purchase one product will decrease the probability of purchase of the other product) and lift value of greater than 1 indicates purchase of Product X will increase the probability of purchase of Product Y.

Lift value of greater than 1 is a necessary condition of generating association rules.

In [4]:
all_txns = []
#open the file
with open('Datasets\\groceries.csv') as f:
    #read each line
    content = f.readlines()
    #Remove white space from the beginning and end of the line
    txns = [x.strip() for x in content]
    # Iterate through each line and create a list of transactions
    for each_txn in txns:
        #Each transaction will contain a list of item in the transaction
        all_txns.append(each_txn.split(',') )

In [5]:
all_txns

[['citrus fruit', 'semi-finished bread', 'margarine', 'ready soups'],
 ['tropical fruit', 'yogurt', 'coffee'],
 ['whole milk'],
 ['pip fruit', 'yogurt', 'cream cheese ', 'meat spreads'],
 ['other vegetables',
  'whole milk',
  'condensed milk',
  'long life bakery product'],
 ['whole milk', 'butter', 'yogurt', 'rice', 'abrasive cleaner'],
 ['rolls/buns'],
 ['other vegetables',
  'UHT-milk',
  'rolls/buns',
  'bottled beer',
  'liquor (appetizer)'],
 ['pot plants'],
 ['whole milk', 'cereals'],
 ['tropical fruit',
  'other vegetables',
  'white bread',
  'bottled water',
  'chocolate'],
 ['citrus fruit',
  'tropical fruit',
  'whole milk',
  'butter',
  'curd',
  'yogurt',
  'flour',
  'bottled water',
  'dishes'],
 ['beef'],
 ['frankfurter', 'rolls/buns', 'soda'],
 ['chicken', 'tropical fruit'],
 ['butter', 'sugar', 'fruit/vegetable juice', 'newspapers'],
 ['fruit/vegetable juice'],
 ['packaged fruit/vegetables'],
 ['chocolate'],
 ['specialty bar'],
 ['other vegetables'],
 ['butter milk

Python library mlxtend provides methods to generate association rules from a list of transactions. But it requires data to be fed in specific format.
The transactions and items need to be converted into a tabular or matrix format. 
Each row represents a transaction and each column represents an item. So, the matrix size will be of M × N, where M represents the total number of transactions and N represents all unique items available across all transactions (or the number of items sold by the seller).

The items available in each transaction will be represented in one-hot-encoded format, that is, the item is encoded 1 if it exists in the transaction or 0 otherwise. 

The mlxtend library has a feature pre-processing implementation class called OnehotTransactions that will take all_txns as an input and convert the transactions and items into one-hot-encoded format

In [8]:
# Import all required libraries
import pandas as pd
import numpy as np
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

In [9]:
# Initialize OnehotTransactions
tx_encoder = TransactionEncoder()
# Transform the data into one-hot-encoding format
txns = tx_encoder.fit(all_txns).transform(all_txns)
# Convert the matrix into the dataframe.
txns_df = pd.DataFrame(txns,columns=tx_encoder.columns_)

In [11]:
txns_df= txns_df.astype(int)

In [12]:
txns_df.head()

Instant food products  UHT-milk  abrasive cleaner  artif. sweetener  \
0                      0         0                 0                 0   
1                      0         0                 0                 0   
2                      0         0                 0                 0   
3                      0         0                 0                 0   
4                      0         0                 0                 0   

   baby cosmetics  baby food  bags  baking powder  bathroom cleaner  beef  \
0               0          0     0              0                 0     0   
1               0          0     0              0                 0     0   
2               0          0     0              0                 0     0   
3               0          0     0              0                 0     0   
4               0          0     0              0                 0     0   

   ...  turkey  vinegar  waffles  whipped/sour cream  whisky  white bread  \
0  ...       0        0        0                   0       0            0   
1  ...       0        0        0                   0       0            0   
2  ...       0        0        0                   0       0            0   
3  ...       0        0        0                   0       0            0   
4  ...       0        0        0                   0       0            0   

   white wine  whole milk  yogurt  zwieback  
0           0           0       0         0  
1           0           0       1         0  
2           0           1       0         0  
3           0           0       1         0  
4           0           1       0         0  

[5 rows x 171 columns]

All items that do not have the minimum support will be removed from the possible itemset combinations.

Apriori algorithm takes the following parameters:
1. df: pandas − DataFrame in a one-hot-encoded format.
2. min_support: A float between 0 and 1 for minimum support of the itemsets returned. Default is 0.5.
3. use_colnames: If true, uses the DataFrames’ column names in the returned DataFrame instead of column indices.

In [26]:
frequent_itemsets = apriori(txns_df,min_support=0.02,use_colnames=True)

C:\Users\tusha\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\mlxtend\frequent_patterns\fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


In [27]:
frequent_itemsets.sample(10, random_state = 90)

support                            itemsets
60   0.020437          (whole milk, bottled beer)
52   0.033859                             (sugar)
89   0.035892  (tropical fruit, other vegetables)
105  0.021047   (tropical fruit, root vegetables)
88   0.032740            (soda, other vegetables)
16   0.058058                            (coffee)
111  0.024504         (whole milk, shopping bags)
36   0.079817                        (newspapers)
119  0.056024                (whole milk, yogurt)
55   0.071683                (whipped/sour cream)

we can infer that whole milk and yogurt appear together in about 5.6% of the baskets.

These itemsets can be passed to association_rules for generating rules and corresponding metrics. 
1. df : DataFrame of frequent itemsets with columns [‘support’, ‘itemsets’].
2. metric − In this use ‘confidence’ and ‘lift’ to evaluate if a rule is of interest. Default is ‘confidence’.
3. min_threshold − Minimal threshold for the evaluation metric to decide whether a candidate rule is of interest.

In [28]:
rules = association_rules(frequent_itemsets, # itemsets
                            metric='lift', # lift
                            min_threshold=1)

In [29]:
rules.sample(5)

antecedents                 consequents  antecedent support  \
123        (whole milk)  (yogurt, other vegetables)            0.255516   
9       (bottled water)                (whole milk)            0.110524   
51   (other vegetables)           (root vegetables)            0.193493   
85         (rolls/buns)                (whole milk)            0.183935   
73          (pip fruit)                (whole milk)            0.075648   

     consequent support   support  confidence      lift  leverage  conviction  \
123            0.043416  0.022267    0.087147  2.007235  0.011174    1.047905   
9              0.255516  0.034367    0.310948  1.216940  0.006126    1.080446   
51             0.108998  0.047382    0.244877  2.246605  0.026291    1.179941   
85             0.255516  0.056634    0.307905  1.205032  0.009636    1.075696   
73             0.255516  0.030097    0.397849  1.557043  0.010767    1.236375   

     zhangs_metric  
123       0.674027  
9         0.200417  
51        0.688008  
85        0.208496  
73        0.387036

In [30]:
rules.sort_values('confidence',ascending = False)

antecedents                 consequents  \
122           (yogurt, other vegetables)                (whole milk)   
17                              (butter)                (whole milk)   
24                                (curd)                (whole milk)   
116  (root vegetables, other vegetables)                (whole milk)   
114        (whole milk, root vegetables)          (other vegetables)   
..                                   ...                         ...   
123                         (whole milk)  (yogurt, other vegetables)   
75                          (whole milk)                      (pork)   
0                           (whole milk)                      (beef)   
30                          (whole milk)               (frankfurter)   
33                          (whole milk)         (frozen vegetables)   

     antecedent support  consequent support   support  confidence      lift  \
122            0.043416            0.255516  0.022267    0.512881  2.007235   
17             0.055414            0.255516  0.027555    0.497248  1.946053   
24             0.053279            0.255516  0.026131    0.490458  1.919481   
116            0.047382            0.255516  0.023183    0.489270  1.914833   
114            0.048907            0.193493  0.023183    0.474012  2.449770   
..                  ...                 ...       ...         ...       ...   
123            0.255516            0.043416  0.022267    0.087147  2.007235   
75             0.255516            0.057651  0.022166    0.086749  1.504719   
0              0.255516            0.052466  0.021251    0.083168  1.585180   
30             0.255516            0.058973  0.020539    0.080382  1.363029   
33             0.255516            0.048094  0.020437    0.079984  1.663094   

     leverage  conviction  zhangs_metric  
122  0.011174    1.528340       0.524577  
17   0.013395    1.480817       0.514659  
24   0.012517    1.461085       0.505984  
116  0.011076    1.457687       0.501524  
114  0.013719    1.533320       0.622230  
..        ...         ...            ...  
123  0.011174    1.047905       0.674027  
75   0.007435    1.031862       0.450546  
0    0.007845    1.033487       0.495856  
30   0.005470    1.023280       0.357751  
33   0.008149    1.034663       0.535554  

[126 rows x 10 columns]

we can infer that the probability that a customer buys (whole milk), given he/she has bought (yogurt, other vegetables), is 0.51.

These rules can be used to create strategies like keeping
the items together in store shelves or cross-selling

Pros and Cons of Association Rule Mining

Transactions data, which is used for generating rules, is always available and mostly clean. 
However, association rules do not take the preference or ratings given by customers into account, which is an important information pertaining for generating rules.

2. COLLABORATIVE FILTERING

Collaborative filtering is based on the notion of similarity (or distance). For example, if two users A and B have purchased the same products and have rated them similarly on a common rating scale, then A and B can be considered similar in their buying and preference behavior. 

Hence, if A buys a new product and rates high, then that product can be recommended to B. Alternatively, the products that A has already bought and rated high can be recommended to B, if not already bought by B.

1. User-Based Similarity: Finds K similar users based on common items they have bought.
2. Item-Based Similarity: Finds K similar items based on common users who have bought those items.


In [31]:
rating_df = pd.read_csv( 'Datasets\\ratings.csv' )
rating_df.head()

userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931

In [32]:
rating_df.drop('timestamp',axis=1,inplace=True)
rating_df.shape

(100836, 3)

In [33]:
(len( rating_df.userId.unique()),len( rating_df.movieId.unique()))

(610, 9724)

We need to create a pivot table or matrix and represent users as rows and
movies as columns.So, a matrix of size 610 X 9724

1. index: Column value to be used as DataFrame’s index. So, it will be userId column of rating_df.
2. columns: Column values to be used as DataFrame’s columns. So, it will be movieId column of rating_df.
3. values: Column to use for populating DataFrame’s values. So, it will be rating column of rating_df

In [34]:
user_movies_df = rating_df.pivot( index='userId',columns='movieId',
                                 values = 'rating').reset_index(drop=True)
user_movies_df.index=rating_df.userId.unique()

In [37]:
user_movies_df.iloc[0:5,0:10]

movieId   1   2    3   4   5    6   7   8   9   10
1        4.0 NaN  4.0 NaN NaN  4.0 NaN NaN NaN NaN
2        NaN NaN  NaN NaN NaN  NaN NaN NaN NaN NaN
3        NaN NaN  NaN NaN NaN  NaN NaN NaN NaN NaN
4        NaN NaN  NaN NaN NaN  NaN NaN NaN NaN NaN
5        4.0 NaN  NaN NaN NaN  NaN NaN NaN NaN NaN

In [38]:
user_movies_df.fillna( 0, inplace = True )
user_movies_df.iloc[0:5, 0:10]

movieId   1    2    3    4    5    6    7    8    9    10
1        4.0  0.0  4.0  0.0  0.0  4.0  0.0  0.0  0.0  0.0
2        0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
3        0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
4        0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
5        4.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0

Each row in user_movies_df represents a user. If we compute the similarity between rows, it will represent the similarity between those users. 
sklearn.metrics.pairwise_distances can be used to compute distance
between all pairs of users.

In [39]:
from sklearn.metrics import pairwise_distances
from scipy.spatial.distance import cosine, correlation
user_sim = 1 - pairwise_distances( user_movies_df.values, metric='cosine' )
#Store the results in a dataframe
user_sim_df = pd.DataFrame( user_sim )
#Set the index and column names to user ids (0 to 671)
user_sim_df.index = rating_df.userId.unique()
user_sim_df.columns = rating_df.userId.unique()

In [40]:
user_sim_df.head()

1         2         3         4         5         6         7    \
1  1.000000  0.027283  0.059720  0.194395  0.129080  0.128152  0.158744   
2  0.027283  1.000000  0.000000  0.003726  0.016614  0.025333  0.027585   
3  0.059720  0.000000  1.000000  0.002251  0.005020  0.003936  0.000000   
4  0.194395  0.003726  0.002251  1.000000  0.128659  0.088491  0.115120   
5  0.129080  0.016614  0.005020  0.128659  1.000000  0.300349  0.108342   

        8         9         10   ...       601       602       603       604  \
1  0.136968  0.064263  0.016875  ...  0.080554  0.164455  0.221486  0.070669   
2  0.027257  0.000000  0.067445  ...  0.202671  0.016866  0.011997  0.000000   
3  0.004941  0.000000  0.000000  ...  0.005048  0.004892  0.024992  0.000000   
4  0.062969  0.011361  0.031163  ...  0.085938  0.128273  0.307973  0.052985   
5  0.429075  0.000000  0.030611  ...  0.068048  0.418747  0.110148  0.258773   

        605       606       607       608       609       610  
1  0.153625  0.164191  0.269389  0.291097  0.093572  0.145321  
2  0.000000  0.028429  0.012948  0.046211  0.027565  0.102427  
3  0.010694  0.012993  0.019247  0.021128  0.000000  0.032119  
4  0.084584  0.200395  0.131746  0.149858  0.032198  0.107683  
5  0.148758  0.106435  0.152866  0.135535  0.261232  0.060792  

[5 rows x 610 columns]

user_sim_df matrix shape shows that it contains the cosine similarity between all possible pairs of users.
And each cell represents the cosine similarity between two specific users.

The diagonal of the matrix shows the similarity of an user with itself (i.e., 1.0). This is true as each user is most similar to himself or herself. But we need the algorithm to find other users who are similar to a specific user. So, we will set the diagonal values as 0.0 

In [41]:
np.fill_diagonal( user_sim, 0 )
user_sim_df.iloc[0:5, 0:5]

1         2         3         4         5
1  0.000000  0.027283  0.059720  0.194395  0.129080
2  0.027283  0.000000  0.000000  0.003726  0.016614
3  0.059720  0.000000  0.000000  0.002251  0.005020
4  0.194395  0.003726  0.002251  0.000000  0.128659
5  0.129080  0.016614  0.005020  0.128659  0.000000

Filtering Similar Users :

To find most similar users, the maximum values of each column can be filtered.

In [42]:
user_sim_df.idxmax(axis=1)[0:5]


1    266
2    366
3    313
4    391
5    470
dtype: int64

To dive a little deeper to understand the similarity, let us print the similarity values between user 2 and users ranging from 360 to 370.

In [43]:
user_sim_df.iloc[1:2, 360:370]

361       362       363  364       365       366       367       368  \
2  0.012776  0.115081  0.084261  0.0  0.149578  0.300074  0.031699  0.008637   

        369       370  
2  0.016431  0.034816

The output shows that the cosine similarity between userid 2 and userid 366 is 0.300074 and highest.
But why is user 366 most similar to user 2? This can be explained intuitively if we can verify that the two users have watched several movies in common and rated very similarly.

For this, we need to read movies dataset, which contains the movie id along with the movie name.

In [47]:
movies_df = pd.read_csv( 'Datasets\\movies.csv')
movies_df.tail()

movieId                                      title  \
9737   193581  Black Butler: Book of the Atlantic (2017)   
9738   193583               No Game No Life: Zero (2017)   
9739   193585                               Flint (2017)   
9740   193587        Bungo Stray Dogs: Dead Apple (2018)   
9741   193609        Andrew Dice Clay: Dice Rules (1991)   

                               genres  
9737  Action|Animation|Comedy|Fantasy  
9738         Animation|Comedy|Fantasy  
9739                            Drama  
9740                 Action|Animation  
9741                           Comedy

In [50]:
movies_df.drop( 'genres', axis = 1, inplace = True )

In [51]:
def get_user_similar_movies( user1, user2 ):
    # Inner join between movies watched between two users will give
    # the common movies watched.
    common_movies = rating_df[rating_df.userId == user1].merge(
        rating_df[rating_df.userId == user2],on = 'movieId',how = 'inner')
    # join the above result set with movies details
    #return common_movies.merge( movies_df, on = 'movieId' )
    return common_movies.merge( movies_df, on = 'movieId')

In [52]:
common_movies = get_user_similar_movies( 2, 366 )
common_movies

userId_x  movieId  rating_x  userId_y  rating_y  \
0         2     3578       4.0       366       4.5   
1         2     6874       4.0       366       4.0   
2         2    48516       4.0       366       4.5   
3         2    58559       4.5       366       4.0   
4         2    68157       4.5       366       4.5   
5         2    79132       4.0       366       4.0   
6         2    91529       3.5       366       4.0   
7         2   109487       3.0       366       5.0   
8         2   122882       5.0       366       2.0   

                           title  
0               Gladiator (2000)  
1       Kill Bill: Vol. 1 (2003)  
2           Departed, The (2006)  
3        Dark Knight, The (2008)  
4    Inglourious Basterds (2009)  
5               Inception (2010)  
6  Dark Knight Rises, The (2012)  
7            Interstellar (2014)  
8      Mad Max: Fury Road (2015)

In [53]:
common_movies[(common_movies.rating_x >= 4.0) & 
              ((common_movies.rating_y >= 4.0))]

userId_x  movieId  rating_x  userId_y  rating_y  \
0         2     3578       4.0       366       4.5   
1         2     6874       4.0       366       4.0   
2         2    48516       4.0       366       4.5   
3         2    58559       4.5       366       4.0   
4         2    68157       4.5       366       4.5   
5         2    79132       4.0       366       4.0   

                         title  
0             Gladiator (2000)  
1     Kill Bill: Vol. 1 (2003)  
2         Departed, The (2006)  
3      Dark Knight, The (2008)  
4  Inglourious Basterds (2009)  
5             Inception (2010)

Challenges with User-Based Similarity

Finding user similarity does not work for new users. We need to wait until the new user buys a few items and rates them. Only then users with similar preferences can be found and recommendations can be made based on that. 

This is called cold start problem in recommender systems. This can be overcome by using item-based similarity. Item-based similarity is based on the notion that if two items have been bought by many users and rated similarly, then there must be some inherent relationship between these two items.

In other terms, in future, if a user buys one of those two items, he or she will most likely buy the other one.

USING SURPRISE LIBRARY

For real-world implementations, we need a more extensive library which hides all the implementation details and provides abstract Application Programming Interfaces (APIs) to build recommender systems.

Various ready-to-use prediction algorithms like neighborhood methods (user similarity and item similarity), and matrix factorization-based. It also has built-in similarity measures such as cosine, mean square distance (MSD), Pearson correlation coefficient, etc.

# surprise Import Remaining

In [54]:
from surprise import Dataset, Reader, KNNBasic, evaluate, accuracy

ModuleNotFoundError: No module named 'surprise'